More Descriptive Errors (Part 1/2)

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#260

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have reviewed some error data-types and constructors to make them more consistent with the rest. 
- [x] Added more extensive descriptions and reviewed some errors code to provide a better feedback to users of the API and/or CLI.

# Comments

<!-- Additional comments or screenshots to attach if any -->

Waiting for @jonathanknowles to control the English here :stuck_out_tongue: 

A few examples:

```
{"code":"wallet_already_exists","message":"This operation would yield a wallet with the following id: 23c401a00f94d05e3e48599ed85cc5ea0e16a2b6. However, I already know a wallet with such id!"}

{"code":"no_such_wallet","message":"I couldn't find a wallet with the given id: 2512a00e9653fe49a44a5886202e24d77eeb998f."}
```




<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ commented on test/integration/Test/Integration/Framework/Auxiliary.hs:
```diff
@@ -0,0 +1,24 @@
+
+mkFixedSizedList :: forall n a. KnownNat n => [a] -> Maybe (FixedSizedList n a)
+mkFixedSizedList v | L.length v == l = Just (UnsafeMkList v)
+                   | otherwise       = Nothing
```
Minor things but: no more variable-length indentation please :pray:

KtorZ commented on test/integration/Test/Integration/Scenario/Transactions.hs:
```diff
@@ -344,35 +345,42 @@ spec = do
         :: String
         -> [Word64]
-        -> [Word64]
+        -> Maybe (TwoElementList Word64)
```
This shouldn't be a `Maybe`; this is made fairly obvious by the way the function handles that with a `case` right away, failing hard in the case it receives a `Nothing`. Generally-speaking, make the input domain no more than what you need. 
This is up to the caller to produce a value of the right type.

paweljakubas replied:
yes, the caller should take responsibility here

paweljakubas commented:
thanks for the remarks! I reconsidered the issue and finally opted for 2-tuple (although still think FixedSizedList has the potential to be used in the future but needs more helper functions to be added)

KtorZ committed:
review some error data-types to make them more consistent

- Trying to preserve the pattern where error data-types and constructor
  are prefixed with `Err...` which identifies them right away as errors.

KtorZ committed:
add extensive description to API errors and review some of the error codes

KtorZ committed:
define middleware to adjust raw servant errors before they're sent

KtorZ committed:
implement 'LiftHandler' for 'ServantErr' and adjust raw servant errors

KtorZ committed:
factor out logic for starting wallet server, and mount middleware

KtorZ committed:
convert double-quotes to single quotes in error message to prevent unfriendly escaping issues

jonathanknowles reviewed (commented):
Hi @KtorZ!

I have to admit, I'm not yet convinced that it's a good idea to use the `QuasiText` quasiquoter here. I can think of a few disadvantages:

1. It silently inserts a `\n` at the end of each line. While it's true that our code is wrapped to a width of 80 characters, this particular wrapping might not be appropriate for the user's view port. (Error messages can of course be viewed in different contexts: perhaps we should let the user's view port determine the wrapping?)

2. It doesn't play nicely with our 4-space indentation rule. If we do indent, then extra spaces are inserted in the resultant string. If we don't indent, then it breaks the nested structure of our code and makes the code harder to read (in my opinion).

3. It doesn't play nicely with syntax highlighting. Most Haskell syntax highlighters can highlight string literals in a different colour (e.g., bright red), but that won't work with this quasi-quoter. In fact, certain words such as "data", if included in the string, could mistakenly be highlighted as keywords. This makes it harder to visually differentiate strings from code. ![dodgy-syntax-highlighting](https://user-images.githubusercontent.com/206319/58146972-ffc5e080-7c8a-11e9-8756-bc810e69c0ba.png)

Compare the following two snippets:
```hs
instance LiftHandler ErrCoinSelection where
    handler = \case
        ErrNotEnoughMoney utxo payment ->
            apiError err403 NotEnoughMoney [embed|
I can't accomodate such payment because there's not enough available UTxO in the
wallet. The total UTxO sums up to $(utxo) Lovelace, but I need $(payment)
Lovelace (fee included) in order to proceed with the payment.|]
        ErrUtxoNotEnoughFragmented nUtxo nOuts ->
            apiError err403 UtxoNotEnoughFragmented [embed|
There's a restriction in the way I can construct transactions: I do not re-use a
same UTxO for different outputs. Here, I only have $(nUtxo) available but there
are $(nOuts) outputs.|]
        ErrMaximumInputsReached n ->
            apiError err403 TransactionIsTooBig [embed|
I had to select as many as $(n) inputs to construct the requested transaction.
This would create a transaction that is too big and would be rejected by a core
node. Try sending a smaller amount.|]
```
with:
```hs
instance LiftHandler ErrCoinSelection where 
    handler = \case 
        ErrNotEnoughMoney utxo payment ->   
            apiError err403 NotEnoughMoney $ mconcat
                [ "I can't accommodate this payment because there's not enough "
                , "available UTxO in the wallet. The total UTxO sums up to "
                , toText utxo, " Lovelace, but I need ", toText payment 
                , " Lovelace (fee included) in order to proceed with the "  
                , "payment." ]  
        ErrUtxoNotEnoughFragmented nUtxo nOuts ->   
            apiError err403 UtxoNotEnoughFragmented $ mconcat   
                [ "There's a restriction in the way I can construct "   
                , "transactions: I do not re-use a same UTxO for different "
                , "outputs. Here, I only have ", toText nUtxo, " available but "
                , "there are ", toText nOuts, " outputs." ] 
        ErrMaximumInputsReached n ->
            apiError err403 TransactionIsTooBig $ mconcat   
                [ "I had to select as many as ", toText n, " inputs to "
                , "construct the requested transaction. This would create a "   
                , "transaction that is too big and would be rejected by a core "
                , "node. Try sending a smaller amount." ]
```
In the above pair of examples, there's not much difference in vertical space usage, but the latter has several advantages:
1. There are no `\n`s silently inserted.
2. It's indented according to the logical structure of the code.
3. Syntax highlighting works as normal.

Perhaps I'm missing something?

piotr-iohk commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -266,56 +278,142 @@ class LiftHandler e where
-                \following output address was not found: " <> pretty addr
+        ErrSignTx (ErrKeyNotFoundForAddress addr) ->
+            apiError err403 KeyNotFoundForAddress $ mconcat
+                [ "I couldn't sign the given transaction: I didn't found the "
```
I suppose: _I didn't find_ or _I haven't found_

KtorZ replied:
Fixed :+1:

piotr-iohk reviewed (approved):
lgtm :+1:  (except one small `typo`)
